src/parse.rs

```diff
@@ -0,0 +1,2602 @@
+    }
+}
+
+fn clause_guard_reduction(
+    op: Spanned,
+    l: UntypedClauseGuard,
+    r: UntypedClauseGuard,
+) -> UntypedClauseGuard {
+    let location = SrcSpan {
+        start: l.location().start,
+        end: r.location().end,
+    };
+    let left = Box::new(l);
+    let right = Box::new(r);
+    match op.1 {
+        Tok::VbarVbar => ClauseGuard::Or {
+            location,
+            left,
+            right,
+        },
+
+        Tok::AmperAmper => ClauseGuard::And {
+            location,
+            left,
+            right,
+        },
+
+        Tok::EqualEqual => ClauseGuard::Equals {
+            location,
+            left,
+            right,
+        },
+
+        Tok::NotEqual => ClauseGuard::NotEquals {
+            location,
+            left,
+            right,
+        },
+
+        Tok::Greater => ClauseGuard::GtInt {
+            location,
+            left,
+            right,
+        },
+
+        Tok::GreaterEqual => ClauseGuard::GtEqInt {
+            location,
+            left,
+            right,
+        },
+
+        Tok::Less => ClauseGuard::LtInt {
+            location,
+            left,
+            right,
+        },
+
+        Tok::LessEqual => ClauseGuard::LtEqInt {
+            location,
+            left,
+            right,
+        },
+
+        Tok::GreaterDot => ClauseGuard::GtFloat {
+            location,
+            left,
+            right,
+        },
+
+        Tok::GreaterEqualDot => ClauseGuard::GtEqFloat {
+            location,
+            left,
+            right,
+        },
+
+        Tok::LessDot => ClauseGuard::LtFloat {
+            location,
+            left,
+            right,
+        },
+
+        Tok::LessEqualDot => ClauseGuard::LtEqFloat {
+            location,
+            left,
+            right,
+        },
+
+        _ => panic!(parser_panic_text(
+            "Token could not be converted to Guard Op."
+        )),
+    }
+}
+
+fn parser_panic_text(message: &str) -> String {
+    format!(
+        "UH OH, looks like you've run into a compiler error. \
+Please report it along with a description of what caused it! \
+Reason: {}",
+        message
+    )
+}
+
+// Bitstring Parse Helpers
+//
+// Bitstrings in patterns, guards, and expressions have a very similar structure
+// but need specific types. These are helpers for that. There is probably a
+// rustier way to do this :)
+fn bit_string_pattern_int(value: String, start: usize, end: usize) -> UntypedPattern {
+    Pattern::Int {
+        location: SrcSpan { start, end },
+        value,
+    }
+}
+
+fn bit_string_expr_int(value: String, start: usize, end: usize) -> UntypedExpr {
+    UntypedExpr::Int {
+        location: SrcSpan { start, end },
+        value,
+    }
+}
+
+fn bit_string_const_int(value: String, start: usize, end: usize) -> UntypedConstant {
+    Constant::Int {
+        location: SrcSpan { start, end },
+        value,
+    }
+}
+
+// Question: parse error here?
```

Comment:
Yes I think so. Removing `Invalid` is on my TODO list

Reply:
Now:
```
error: Syntax error
  ┌─ /Users/a/Trash/parser_test/src/one.gleam:6:7
  │
6 │   <<a:invalid>>
  │       ^^^^^^^ This is not a valid BitString segment option.

Hint: Valid BitString segment options are:
binary, int, float, bit_string, utf8, utf16, utf32, utf8_codepoint, utf16_codepoint
utf32_codepoint, signed, unsigned, big, little, native, size, unit
See: https://gleam.run/book/tour/bit-strings
```

Reply:
Much better